### ISSUE TITLE
I think tf.nn.moments() produces negative variance.

### ISSUE THREAD

**Issue by cesc-park** [tensorflow/tensorflow#3290]
tf.nn.normalize_moments() calculates variance like this.

```
variance = math_ops.sub(
        math_ops.mul(variance_ss, divisor),
        math_ops.square(shifted_mean),
        name="variance")
```

This code can cause negative variance due to floating point precision problem.

Thanks.

### Logs or other output that would be helpful
(If logs are large, please upload as attachment).

tf.Print output for variance from tf.nn.moments().
```
I tensorflow/core/kernels/logging_ops.cc:79] variance[-0.0021057129 0.00061035156 -0.00024414062...]
I tensorflow/core/kernels/logging_ops.cc:79] variance[14.041901 5.8100648 60.691284...]
I tensorflow/core/kernels/logging_ops.cc:79] variance[0.46612549 0.017944336 0.87060547...]
```

**Comment by jmchen-g** [comment]
Thanks for reporting. Actually @zheng-xq is already working on this. Any updates XQ?

**Comment by zheng-xq** [comment]
We are aware of this issue caused by numerical precision. Internally we are discussing a solution with a better tradeoff between numerical stability, memory and computation efficiency. 

For now, please work around this issue by clamp the variance to zero, or a small positive value.

**Comment by michaelisard** [comment]
Closing for now. Please reopen if the workaround is not suitable for you.

**Comment by daeyun** [comment]
was this fixed? I'm wondering if this happens with `tensorflow.contrib.slim.batch_norm`.

**Comment by zheng-xq** [comment]
We are switching to a fused batch norm implemenation. The new op is here: 

https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/nn.py#L985

We are in the process of switching the default layers.batch_norm. It will be announced when it is done.

**Comment by eywalker** [comment]
Indeed I'm observing `nan` values from `tensorflow.contrib.slim.batch_norm` due to negative variances returned from `tf.nn.moments`. Looking forward to seeing this addressed.

**Comment by yaroslavvb** [comment]
It's addressed with tf.nn.fused_batch_norm

**Comment by cesc-park** [comment]
@michaelisard this still can produce negative variance.
Why not use the Var = E[square(X- mean)] equation instead of the Var = E[square(X)] - square(mean) equation?

**Comment by yazabaza** [comment]
Any update on this? Caused a mild train wreck here. Perhaps a warning in the documentation might benefit others.

**Comment by yaroslavvb** [comment]
@yazabaza which version? There have been some updates to that functionality, can you reproduce with nightly?

**Comment by cesc-park** [comment]
tf.nn.fused_batch_norm works great, because it use "Var = E[square(X- mean)]" to calculate variance. I think tf.nn.moments should use it too.

**Comment by yazabaza** [comment]
I am running version 0.12.0. Haven't yet tried 1.0 or nightly. I wanted to use tf.nn.fused_batch_norm but could not decipher its documentation. I might just create my own variance calculation with Tensorflow math ops.

**Comment by yazabaza** [comment]
I tested in v1.0.1 and this issue appears to be resolved. Thanks!

**Comment by pawni** [comment]
I'm also dealing with this issue on TF 1.0.1. Furthermore, tf.nn.fused_batch_norm isn't a solution because it requires 4D input and I'm working with 5D tensors and 3D convolutions. Is there any progress on this?

**Comment by raghavgoyal14** [comment]
I'm also getting the negative variance while using `tf.contrib.layers.batch_norm(fused=True)`.
Should I use `tf.nn.fused_batch_norm` ?

**Comment by appyfizzA** [comment]
@pawni Did you find a solution? I'm also working with 3D convolutions and 5D tensors, and I'm currently using tf.nn.moments
